functional_tests/README.md

```diff
@@ -0,0 +1,303 @@
+pass in explicit variables from the context and return new ones as required. Try to make all none step functions
+pure/deterministic.
+
+This is to avoid the context use becoming obscured and hard to follow, when context variable are only set in top level
+step functions then it is easy to follow where the variables for any particular scenario are set, simply by walking
+through the steps. Passing the entire context object down to lower level functions effectively obscures their true
+signatures, making the use and setting of context variables much harder to follow.
+
+For example:
+
+> [!TIP]
+>
+> **Do this:**
+>
+> ```python
+> @step('a thing happens')
+> def step_to_do_a_thing(context: Context):
+>     context.a_new_context_var = helper_function(context.my_scenario_data)
+>     ...
+>
+> def helper_function(my_scenario_data):
+>     ...
+>     return new_data
+> ```
+
+> [!CAUTION]
+>
+> **Not this:**
+>
+> ```python
+> @step('a thing happens')
+> def step_to_do_a_thing(context: Context):
+>     helper_function_which_overwrites_context(context)
+>     ...
+>
+> def helper_function_which_overwrites_context(context: Context) -> None:
+>     ...
+>     context.a_new_context_var = new_data
+> ```
+
+### Sharing Code Between Steps
+
+Step files should not import code from other step files, where code can be shared between steps they should either be in
+the same file, or the shared code should be factored out into the [step_helpers](step_helpers) module.
+
+This is to avoid potential circular imports and make it clear which code is specific to certain steps, and which is
+reusable across any steps.
+
+### Step wording
+
+Steps should be written in full and concise sentences, avoiding unnecessary abbreviations and shorthand. They should be
+as understandable and as non-technical as possible.
+
+### Assertions
+
+Assertions should use the [Playwright assertions](https://playwright.dev/docs/test-assertions) wherever possible. This
+has built in retry and timeout logic where applicable, to allow grace periods for content to load on the page, so will
+be more robust than attempting to retrieve data from the page and make plain python assertions.
+
+### Step parameter types
+
+Where we need [step parameters](https://behave.readthedocs.io/en/stable/tutorial.html#step-parameters) to include more
+complex data than single strings or the other basic types supported by the default parser, we
+use [custom registered types](https://behave.readthedocs.io/en/stable/api.html#behave.register_type). These are
+registered in the [environment.py](environment.py) so they are available to all steps.
+
+## How the Tests Work
+
+### Django Test Runner and Test Case
+
+Due to issues with the Django `TransactionTestCase` which prevent us using the built-in database teardown/setup
+in between scenarios, we have implemented our own database snapshot and restore pattern between tests. We still make use
```

Comment:
Question: What is this issue?

You can use [`serialized_rollback`](https://docs.djangoproject.com/en/5.1/topics/testing/overview/#rollback-emulation) to ensure data created in migrations is loaded between tests.

Or is the need more about having a live server to test against?

Reply:
I believe there were several. off the top of my head:

1. the setup tear down was rather slow (and there was some quirk with behave)
2. The `Locale` issue which is now covered by https://github.com/ONSdigital/dis-wagtail/blob/92ed3b95e77995e28edd923859b23f01fc7b1832/cms/core/tests/__init__.py#L33

Reply:
Yeah the performance isn't something which I've solved - Django warns that the slowdown can be as much as 3x with `serialized_rollback`. But slightly slower tests may be worth the significantly reduced complexity.

Reply:
My preference is to get this in, then revisit. 
Worth keeping in mind that the slowness will increase as we add tests.

Reply:
Agreed, let's revisit if it becomes a problem.